ensime_shared/editor.py

```diff
@@ -0,0 +1,384 @@
+            'autocmd FileType package_info nnoremap <buffer> <Space> :call EnPackageDecl()<CR>')
+        self._vim.command('autocmd FileType package_info setlocal splitright')
+
+    # TODO: make this a R/W property?
+    def cursor(self):
+        """Get the cursor position in the current window as a ``(row, column)``
+        tuple.
+        """
+        return self._vim.current.window.cursor
+
+    def set_cursor(self, row, col):
+        """Set cursor position to given row and column in the current window."""
+        self._vim.current.window.cursor = (row, col)
+
+    # TODO: this can be replaced with a ``getpos()`` expression
+    def start_end_pos(self):
+        """Return start and end positions of the cursor respectively."""
+        self._vim.command('normal e')
+        end = self.cursor()
+        self._vim.command('normal b')
+        beg = self.cursor()
+        return beg, end
+
+    def path(self):
+        """Return the current path."""
+        return self._vim.current.buffer.name
+
+    def get_file_content(self):
+        """Get content of file."""
+        return "\n".join(self._vim.current.buffer)
+
+    # This is used only once, maybe just make a higher-level API or inline it
+    def width(self):
+        """Return the width of the window."""
+        return self._vim.current.window.width
+
+    def ask_input(self, prompt):
+        """Prompt user for input and return the entered value."""
+        self._vim.command('call inputsave()')
+        self._vim.command('let user_input = input("{} ")'.format(prompt))
+        self._vim.command('call inputrestore()')
+        response = self._vim.eval('user_input')
+        self._vim.command('unlet user_input')
+        return response
+
+    def to_quickfix_item(self, file_name, line_number, message, tpe):
+        return {"filename": file_name,
+                "lnum": line_number,
+                "text": message,
+                "type": tpe}
+
+    def write_quickfix_list(self, qflist):
+        self._vim.command('call setqflist({!s})'.format(qflist))
+        self._vim.command('copen')
+
+    def lazy_display_error(self, filename):
+        """Display error when user is over it."""
+        position = self.cursor()
+        error = self.get_error_at(position)
+        if error:
+            report = error.get_truncated_message(position, self.width() - 1)
+            self.raw_message(report)
+
+    def get_error_at(self, cursor):
+        """Return error at position `cursor`."""
+        for error in self._errors:
+            if error.includes(self._vim.eval("expand('%:p')"), cursor):
+                return error
+        return None
+
+    def clean_errors(self):
+        """Clean errors and unhighlight them in vim."""
+        self._vim.eval('clearmatches()')
+        self._errors = []
+        self._matches = []
+        # Reset Syntastic notes - TODO: bufdo?
+        self._vim.current.buffer.vars['ensime_scala_notes'] = []
+
+    def message(self, key):
```

Comment:
Yeah most definitely want to clean this up, but it falls under "direct port" for this review. I know I've created some new API already but I believe it's all stuff that wasn't encapsulated in a function already.